src/gadget/ecc.rs

```diff
@@ -0,0 +1,146 @@
+//! Gadgets and chips for elliptic curve operations.
+
+use std::fmt;
+
+use crate::{
+    arithmetic::CurveAffine,
+    circuit::{Chip, Layouter},
+    plonk::Error,
+};
+
+/// The set of circuit instructions required to use the ECC gadgets.
+pub trait EccInstructions<C: CurveAffine>: Chip<Field = C::Base> {
+    /// Variable representing an element of the elliptic curve's scalar field.
+    type Scalar: Clone + fmt::Debug;
+    /// Variable representing an elliptic curve point.
+    type Point: Clone + fmt::Debug;
+    /// Variable representing a fixed elliptic curve point (constant in the circuit).
+    type FixedPoint: Clone + fmt::Debug;
+
+    /// Witnesses the given scalar as a private input to the circuit.
+    fn witness_scalar(
+        layouter: &mut impl Layouter<Self>,
+        value: impl FnMut() -> Result<C::Scalar, Error>,
```

Comment:
We don't actually want this to be a function; we need to optionally map over the value if it exists, and with a function we can only learn this by calling the function, making it redundant.
```suggestion
        value: Option<C::Scalar>,
```